File: rcon4j-core/src/main/java/moe/mukjep/rcon4j/RconPacket.java
package moe.mukjep.rcon4j;

import moe.mukjep.rcon4j.exceptions.MalformedPacketException;

import java.io.DataInputStream;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.BufferUnderflowException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

public final class RconPacket {
    private final int id;
    private final int type;
    private final byte[] payload;

    public RconPacket(int requestId, int type, byte[] payload) {
        this.id = requestId;
        this.type = type;
        this.payload = payload;
    }

    public byte[] getPayload() {
        return payload;
    }

    public int getType() {
        return type;
    }

    public int getId() {
        return id;
    }

    @Override
    public String toString() {
        return "RconPacket{" +
                "id=" + id +
                ", type=" + type +
                ", payload=" + Arrays.toString(payload) +
                '}';
    }

    public byte[] toBytes() {
        return createPacket(id, type, payload);
    }

    public static byte[] createPacket(int id, int type, byte[] payload) {
        int bodyLength = 4 + 4 + payload.length + 2;
        int packetLength = 4 + bodyLength;

        ByteBuffer buffer = ByteBuffer.allocate(packetLength).order(ByteOrder.LITTLE_ENDIAN);

        buffer.putInt(bodyLength);
        buffer.putInt(id);
        buffer.putInt(type);
        buffer.put(payload);
        buffer.put((byte) 0);
        buffer.put((byte) 0);

        return buffer.array();
    }

    public static RconPacket fromStream(InputStream stream) throws IOException {
        try {
            byte[] header = new byte[4 * 3];
            stream.read(header);

            ByteBuffer buffer = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN);

            int length = buffer.getInt();
            int id = buffer.getInt();
            int type = buffer.getInt();

            byte[] payload = new byte[length - 4 - 4 - 2];
            DataInputStream dis = new DataInputStream(stream);

            dis.readFully(payload);
            dis.readByte();
            dis.readByte();

            return new RconPacket(id, type, payload);
        } catch (BufferUnderflowException | EOFException e) {
            throw new MalformedPacketException("Packet read error");
        }
    }
}

File: rcon4j-core/src/main/kotlin/moe/mukjep/rcon4j/RconPacket.kt
package moe.mukjep.rcon4j

import moe.mukjep.rcon4j.exceptions.MalformedPacketException
import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.nio.BufferUnderflowException
import java.nio.ByteBuffer
import java.nio.ByteOrder

class RconPacket(val id: Int, val type: Int, val payload: ByteArray) {

    @OptIn(ExperimentalStdlibApi::class)
    override fun toString(): String {
        return "RconPacket{" +
                "id=" + id +
                ", type=" + type +
                ", payload=" + payload.toHexString() +
                '}'
    }

    fun toBytes(): ByteArray {
        return bytes(id, type, payload)
    }

    companion object {
        @JvmStatic
        fun bytes(id: Int, type: Int, payload: ByteArray): ByteArray {
            val bodyLength = 4 + 4 + payload.size + 2
            val packetLength = 4 + bodyLength

            val buffer = ByteBuffer.allocate(packetLength).order(ByteOrder.LITTLE_ENDIAN)

            buffer.putInt(bodyLength)
            buffer.putInt(id)
            buffer.putInt(type)
            buffer.put(payload)
            buffer.put(0.toByte())
            buffer.put(0.toByte())

            return buffer.array()
        }

        @JvmStatic
        @JvmName("from")
        @Throws(IOException::class)
        fun InputStream.readRconPacket(): RconPacket {
            try {
                val header = ByteArray(4 * 3)
                read(header)

                val buffer = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN)

                val length = buffer.getInt()
                val id = buffer.getInt()
                val type = buffer.getInt()

                val payload = ByteArray(length - 4 - 4 - 2)

                read(payload)
                read()
                read()

                return RconPacket(id, type, payload)
            } catch (e: BufferUnderflowException) {
                throw MalformedPacketException("Packet read error")
            } catch (e: EOFException) {
                throw MalformedPacketException("Packet read error")
            }
        }
    }
}
